File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/UserProfileFragment.java
package org.mifos.mobilebanking.ui.fragments;

import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.api.local.PreferencesHelper;
import org.mifos.mobilebanking.models.client.Client;
import org.mifos.mobilebanking.models.client.Group;
import org.mifos.mobilebanking.presenters.UserDetailsPresenter;
import org.mifos.mobilebanking.ui.activities.EditUserDetailActivity;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.UserDetailsView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.DateHelper;
import org.mifos.mobilebanking.utils.TextDrawable;
import org.mifos.mobilebanking.utils.Toaster;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler;

/**
 * Created by dilpreet on 10/7/17.
 */

public class UserProfileFragment extends BaseFragment implements UserDetailsView {

    @BindView(R.id.iv_profile)
    ImageView ivProfile;

    @BindView(R.id.iv_text_drawable)
    ImageView ivTextDrawable;

    @BindView(R.id.app_bar_layout)
    AppBarLayout appBarLayout;

    @BindView(R.id.collapsing_toolbar)
    CollapsingToolbarLayout collapsingToolbarLayout;

    @BindView(R.id.tv_user_name)
    TextView tvUsername;

    @BindView(R.id.tv_account_number)
    TextView tvAccountNumber;

    @BindView(R.id.tv_activation_date)
    TextView tvActivationDate;

    @BindView(R.id.tv_office_name)
    TextView tvOfficeName;

    @BindView(R.id.tv_groups)
    TextView tvGroups;

    @BindView(R.id.tv_client_type)
    TextView tvClientType;

    @BindView(R.id.tv_client_classification)
    TextView tvClientClassification;

    @BindView(R.id.tv_phone_number)
    TextView tvPhoneNumber;

    @BindView(R.id.tv_dob)
    TextView tvDOB;

    @BindView(R.id.tv_gender)
    TextView tvGender;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.layout_error)
    View layoutError;

    @BindView(R.id.ll_user_profile)
    LinearLayout llUserProfile;

    @BindView(R.id.fab_edit)
    FloatingActionButton fabEdit;

    @Inject
    UserDetailsPresenter presenter;

    @Inject
    PreferencesHelper preferencesHelper;

    private View rootView;
    private Bitmap userBitmap;
    private Client client;
    private SweetUIErrorHandler sweetUIErrorHandler;

    public static UserProfileFragment newInstance() {
        UserProfileFragment fragment = new UserProfileFragment();
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_user_profile, container, false);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        ButterKnife.bind(this, rootView);
        presenter.attachView(this);

        ((BaseActivity) getActivity()).setSupportActionBar(toolbar);
        ((BaseActivity) getActivity()).getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        collapsingToolbarLayout.setCollapsedTitleTextColor(ContextCompat.getColor(getActivity(),
                R.color.white));
        collapsingToolbarLayout.setExpandedTitleColor(ContextCompat.getColor(getActivity(),
                                R.color.white));

        sweetUIErrorHandler = new SweetUIErrorHandler(getActivity(), rootView);
        if (savedInstanceState == null) {
            presenter.getUserDetails();
            presenter.getUserImage();
        }
        return rootView;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(Constants.USER_DETAILS, client);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (savedInstanceState != null) {
            client = savedInstanceState.getParcelable(Constants.USER_DETAILS);
            presenter.setUserProfile(preferencesHelper.getUserProfileImage());
            showUserDetails(client);
        }
    }

    /**
     * Sets client basic details which are fetched from server
     *
     * @param client instance of {@link Client} which contains information about client
     */
    @Override
    public void showUserDetails(Client client) {
        this.client = client;
        tvUsername.setText(nullFieldCheck(getString(R.string.username), client.getDisplayName()));
        tvAccountNumber.setText(nullFieldCheck(getString(R.string.account_number),
                client.getAccountNo()));
        tvActivationDate.setText(nullFieldCheck(getString(R.string.activation_date),
                DateHelper.getDateAsString(client.getActivationDate())));
        tvOfficeName.setText(nullFieldCheck(getString(R.string.office_name),
                client.getOfficeName()));
        tvClientType.setText(nullFieldCheck(getString(R.string.client_type),
                client.getClientType().getName()));
        tvGroups.setText(nullFieldCheck(getString(R.string.groups),
                getGroups(client.getGroups())));
        tvClientClassification.setText(nullFieldCheck(getString(R.string.client_classification),
                client.getClientClassification().getName()));
        tvPhoneNumber.setText(nullFieldCheck(getString(R.string.phone_number),
                client.getMobileNo()));
        if (client.getDobDate().size() != 3) {  // no data entry in database for the client
            tvDOB.setText(getString(R.string.no_dob_found));
        } else {
            tvDOB.setText(DateHelper.getDateAsString(client.getDobDate()));
        }
        tvGender.setText(nullFieldCheck(getString(R.string.gender), client.getGender().getName()));
    }

    private String nullFieldCheck(String field, String value) {
        if (value == null) {
            return getString(R.string.no) + getString(R.string.blank) + field +
                    getString(R.string.blank) + getString(R.string.found);
        }
        return value;
    }

    /**
     * Generate a string for groups which the client is part of.
     *
     * @param groups {@link List} of {@link Group} which client is a part of.
     * @return Returns String of groups
     */
    private String getGroups(List<Group> groups) {
        if (groups.size() == 0) {
            return getString(
                    R.string.not_assigned_with_any_group); // no groups entry in database for the
            // client
        }


        StringBuilder builder = new StringBuilder();
        for (Group group : groups) {
            builder.append(getString(R.string.string_and_string, group.getName(), " | "));
        }
        return builder.toString().substring(0, builder.toString().length() - 2);
    }

    /**
     * Provides with client Image fetched from the server in {@code bitmap}
     *
     * @param bitmap User Image
     */
    @Override
    public void showUserImage(final Bitmap bitmap) {
        getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                userBitmap = bitmap;
                if (userBitmap == null) {
                    final TextDrawable textDrawable = TextDrawable.builder()
                            .beginConfig()
                            .toUpperCase()
                            .endConfig()
                            .buildRound(preferencesHelper
                                    .getClientName()
                                    .substring(0, 1),
                                    ContextCompat.getColor(getContext(), R.color.primary_dark));
                    ivProfile.setVisibility(View.GONE);
                    ivTextDrawable.setVisibility(View.VISIBLE);
                    ivTextDrawable.setImageDrawable(textDrawable);
                } else {
                    ivTextDrawable.setVisibility(View.GONE);
                    ivProfile.setVisibility(View.VISIBLE);
                    ivProfile.setImageBitmap(bitmap);
                }
            }
        });
    }

    @OnClick(R.id.btn_change_password)
    void changePassword() {
        startActivity(new Intent(getContext(), EditUserDetailActivity.class));
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param message Error message that tells the user about the problem.
     */
    @Override
    public void showError(String message) {
        Toaster.show(rootView, message);
        sweetUIErrorHandler.showSweetCustomErrorUI(getString(R.string.error_fetching_user_profile),
                R.drawable.ic_assignment_turned_in_black_24dp, appBarLayout,
                layoutError);
        fabEdit.setVisibility(View.GONE);
    }

    @Override
    public void showProgress() {
        showProgressBar();
    }

    @Override
    public void hideProgress() {
        hideProgressBar();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        hideProgress();
        presenter.detachView();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/UserProfileFragment.kt
package org.mifos.mobile.ui.fragments

import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobile.R
import org.mifos.mobile.api.local.PreferencesHelper
import org.mifos.mobile.databinding.FragmentUserProfileBinding
import org.mifos.mobile.models.client.Client
import org.mifos.mobile.models.client.Group
import org.mifos.mobile.ui.activities.EditUserDetailActivity
import org.mifos.mobile.ui.activities.base.BaseActivity
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.ui.getThemeAttributeColor
import org.mifos.mobile.utils.*
import org.mifos.mobile.viewModels.UserDetailViewModel
import javax.inject.Inject

/**
 * Created by dilpreet on 10/7/17.
 */
@AndroidEntryPoint
class UserProfileFragment : BaseFragment() {

    private var _binding: FragmentUserProfileBinding? = null
    private val binding get() = _binding!!

    private lateinit var viewModel: UserDetailViewModel

    @JvmField
    @Inject
    var preferencesHelper: PreferencesHelper? = null
    private var userBitmap: Bitmap? = null
    private var client: Client? = null
    private var sweetUIErrorHandler: SweetUIErrorHandler? = null
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentUserProfileBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(this)[UserDetailViewModel::class.java]
        (activity as BaseActivity?)?.setSupportActionBar(binding.toolbar)
        (activity as BaseActivity?)?.supportActionBar?.setDisplayHomeAsUpEnabled(true)
        sweetUIErrorHandler = SweetUIErrorHandler(activity, binding.root)
        if (savedInstanceState == null) {
            viewModel.userDetails
            viewModel.userImage
        }
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.userDetailUiState.observe(viewLifecycleOwner) {
            when (it) {
                is UserDetailUiState.Loading -> showProgress()
                is UserDetailUiState.ShowUserDetails -> {
                    hideProgress()
                    showUserDetails(it.client)
                }
                is UserDetailUiState.ShowUserImage -> {
                    hideProgress()
                    showUserImage(it.image)
                }
                is UserDetailUiState.ShowError -> {
                    hideProgress()
                    showError(getString(it.message))
                }
            }
        }

        binding.btnChangePassword.setOnClickListener {
            changePassword()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(Constants.USER_DETAILS, client)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState != null) {
            client = savedInstanceState.getParcelable(Constants.USER_DETAILS)
            viewModel.setUserProfile(preferencesHelper?.userProfileImage)
            showUserDetails(client)
        }
    }

    /**
     * Sets client basic details which are fetched from server
     *
     * @param client instance of [Client] which contains information about client
     */
    fun showUserDetails(client: Client?) {
        this.client = client
        with(binding) {
            tvUserName.text = nullFieldCheck(getString(R.string.username), client?.displayName)
            tvAccountNumber.text = nullFieldCheck(
                getString(R.string.account_number),
                client?.accountNo,
            )
            tvActivationDate.text = nullFieldCheck(
                getString(R.string.activation_date),
                DateHelper.getDateAsString(client?.activationDate),
            )
            tvOfficeName.text = nullFieldCheck(
                getString(R.string.office_name),
                client?.officeName,
            )
            tvClientType.text = nullFieldCheck(
                getString(R.string.client_type),
                client?.clientType?.name,
            )
            tvGroups.text = nullFieldCheck(
                getString(R.string.groups),
                getGroups(client?.groups),
            )
            tvClientClassification.text = client?.clientClassification?.name ?: "-"
            tvPhoneNumber.text = nullFieldCheck(
                getString(R.string.phone_number),
                client?.mobileNo,
            )
            if (client?.dobDate?.size != 3) { // no data entry in database for the client
                tvDob.text = getString(R.string.no_dob_found)
            } else {
                tvDob.text = DateHelper.getDateAsString(client.dobDate)
            }
            tvGender.text = nullFieldCheck(getString(R.string.gender), client?.gender?.name)
        }

    }

    private fun nullFieldCheck(field: String, value: String?): String {
        return value
            ?: getString(R.string.no) + getString(R.string.blank) + field +
            getString(R.string.blank) + getString(R.string.found)
    }

    /**
     * Generate a string for groups which the client is part of.
     *
     * @param groups [List] of [Group] which client is a part of.
     * @return Returns String of groups
     */
    private fun getGroups(groups: List<Group>?): String {
        if (groups?.isEmpty() == true) {
            return getString(
                R.string.not_assigned_with_any_group,
            ) // no groups entry in database for the
            // client
        }
        val builder = StringBuilder()
        if (groups != null) {
            for ((_, _, name) in groups) {
                builder.append(getString(R.string.string_and_string, name, " | "))
            }
        }
        return builder.toString().substring(0, builder.toString().length - 2)
    }

    /**
     * Provides with client Image fetched from the server in `bitmap`
     *
     * @param bitmap User Image
     */
    fun showUserImage(bitmap: Bitmap?) {
        activity?.runOnUiThread {
            userBitmap = bitmap
            if (userBitmap == null) {
                val textDrawable = TextDrawable.builder()
                    .beginConfig()
                    .toUpperCase()
                    .endConfig()
                    .buildRound(
                        (
                                if (preferencesHelper?.clientName.isNullOrEmpty()) {
                                    preferencesHelper
                                        ?.userName
                                } else {
                                    preferencesHelper
                                        ?.clientName
                                }
                                )
                            ?.substring(0, 1),
                        requireContext().getThemeAttributeColor(R.attr.colorPrimaryVariant),
                    )
                binding.ivProfile.setImageDrawable(textDrawable)
            } else {
                binding.ivProfile.setImageBitmap(bitmap)
            }
        }
    }

    private fun changePassword() {
        startActivity(Intent(context, EditUserDetailActivity::class.java))
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param message Error message that tells the user about the problem.
     */
    fun showError(message: String?) {
        Toaster.show(binding.root, message)
        sweetUIErrorHandler?.showSweetCustomErrorUI(
            getString(R.string.error_fetching_user_profile),
            R.drawable.ic_error_black_24dp,
            binding.appBarLayout,
            binding.layoutError.root,
        )
        binding.fabEdit.visibility = View.GONE
    }

    fun showProgress() {
        showProgressBar()
    }

    fun hideProgress() {
        hideProgressBar()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        hideProgress()
        _binding = null
    }

    companion object {
        @JvmStatic
        fun newInstance(): UserProfileFragment {
            return UserProfileFragment()
        }
    }
}
